cmd/tink-worker/worker/container_manager.go

```diff
@@ -0,0 +1,154 @@
+	loggerIface := ctx.Value(loggingContextKey)
+	if loggerIface == nil {
+		return &m.logger
+	}
+	return loggerIface.(*log.Logger)
+}
+
+// NewContainerManager returns a new container manager.
+func NewContainerManager(logger log.Logger, cli DockerClient, registryDetails RegistryConnDetails) ContainerManager {
+	return &containerManager{logger, cli, registryDetails}
+}
+
+func (m *containerManager) CreateContainer(ctx context.Context, cmd []string, wfID string, action *pb.WorkflowAction, captureLogs, privileged bool) (string, error) {
+	l := m.getLogger(ctx)
+	config := &container.Config{
+		Image:        path.Join(m.registryDetails.Registry, action.GetImage()),
+		AttachStdout: true,
+		AttachStderr: true,
+		Cmd:          cmd,
+		Tty:          true,
+		Env:          action.GetEnvironment(),
+	}
+	if !captureLogs {
+		config.AttachStdout = false
+		config.AttachStderr = false
+		config.Tty = false
+	}
+
+	wfDir := filepath.Join(defaultDataDir, wfID)
+	hostConfig := &container.HostConfig{
+		Privileged: privileged,
+		Binds:      []string{wfDir + ":/workflow"},
+	}
+
+	if pidConfig := action.GetPid(); pidConfig != "" {
+		l.With("pid", pidConfig).Info("creating container")
+		hostConfig.PidMode = container.PidMode(pidConfig)
+	}
+
+	hostConfig.Binds = append(hostConfig.Binds, action.GetVolumes()...)
+	l.With("command", cmd).Info("creating container")
+	resp, err := m.cli.ContainerCreate(ctx, config, hostConfig, nil, nil, action.GetName())
+	if err != nil {
+		return "", errors.Wrap(err, "DOCKER CREATE")
+	}
+	return resp.ID, nil
+}
+
+func (m *containerManager) StartContainer(ctx context.Context, id string) error {
+	m.getLogger(ctx).With("containerID", id).Debug("starting container")
+	return errors.Wrap(m.cli.ContainerStart(ctx, id, types.ContainerStartOptions{}), "DOCKER START")
+}
+
+func (m *containerManager) WaitForContainer(ctx context.Context, id string) (pb.State, error) {
+	// Inspect whether the container is in running state
+	if _, err := m.cli.ContainerInspect(ctx, id); err != nil {
+		return pb.State_STATE_FAILED, nil // nolint:nilerr // error is not nil, but it returns nil
+	}
+
+	// send API call to wait for the container completion
+	wait, errC := m.cli.ContainerWait(ctx, id, container.WaitConditionNotRunning)
+
+	select {
+	case status := <-wait:
+		if status.StatusCode == 0 {
+			return pb.State_STATE_SUCCESS, nil
+		}
+		return pb.State_STATE_FAILED, nil
+	case err := <-errC:
+		return pb.State_STATE_FAILED, err
+	case <-ctx.Done():
+		return pb.State_STATE_TIMEOUT, ctx.Err()
+	}
+}
+
+func (m *containerManager) WaitForFailedContainer(ctx context.Context, id string, failedActionStatus chan pb.State) {
+	l := m.getLogger(ctx)
+	// send API call to wait for the container completion
+	wait, errC := m.cli.ContainerWait(ctx, id, container.WaitConditionNotRunning)
+
+	select {
+	case status := <-wait:
+		if status.StatusCode == 0 {
+			failedActionStatus <- pb.State_STATE_SUCCESS
+			return
```

Comment:
This method is only ever called in a goroutine, but if it isn't, this code will block on a non-buffered channel so I added the return here